// angular-split/esm2022/lib/gutter/split-gutter-exclude-from-drag.directive.mjs
import { Directive, Inject, ElementRef } from '@angular/core';
import { SplitGutterDirective } from './split-gutter.directive';
import { GUTTER_NUM_TOKEN } from './gutter-num-token';
import * as i0 from "@angular/core";
import * as i1 from "./split-gutter.directive";
export class SplitGutterExcludeFromDragDirective {
    constructor(gutterNum, elementRef, gutterDir) {
        this.gutterNum = gutterNum;
        this.elementRef = elementRef;
        this.gutterDir = gutterDir;
    }
    ngOnInit() {
        this.gutterDir.addToMap(this.gutterDir.gutterToExcludeDragElementMap, this.gutterNum, this.elementRef);
    }
    ngOnDestroy() {
        this.gutterDir.removedFromMap(this.gutterDir.gutterToExcludeDragElementMap, this.gutterNum, this.elementRef);
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: SplitGutterExcludeFromDragDirective, deps: [{ token: GUTTER_NUM_TOKEN }, { token: i0.ElementRef }, { token: i1.SplitGutterDirective }], target: i0.ɵɵFactoryTarget.Directive }); }
    /** @nocollapse */ static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.8", type: SplitGutterExcludeFromDragDirective, isStandalone: true, selector: "[asSplitGutterExcludeFromDrag]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: SplitGutterExcludeFromDragDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[asSplitGutterExcludeFromDrag]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [GUTTER_NUM_TOKEN]
                }] }, { type: i0.ElementRef }, { type: i1.SplitGutterDirective }] });
