// angular-split/esm2022/lib/gutter/split-gutter.directive.mjs
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class SplitGutterDirective {
    constructor(template) {
        this.template = template;
        /**
         * The map holds reference to the drag handle elements inside instances
         * of the provided template.
         */
        this.gutterToHandleElementMap = new Map();
        /**
         * The map holds reference to the excluded drag elements inside instances
         * of the provided template.
         */
        this.gutterToExcludeDragElementMap = new Map();
    }
    canStartDragging(originElement, gutterNum) {
        if (this.gutterToExcludeDragElementMap.has(gutterNum)) {
            const isInsideExclude = this.gutterToExcludeDragElementMap
                .get(gutterNum)
                .some((gutterExcludeElement) => gutterExcludeElement.nativeElement.contains(originElement));
            if (isInsideExclude) {
                return false;
            }
        }
        if (this.gutterToHandleElementMap.has(gutterNum)) {
            return this.gutterToHandleElementMap
                .get(gutterNum)
                .some((gutterHandleElement) => gutterHandleElement.nativeElement.contains(originElement));
        }
        return true;
    }
    addToMap(map, gutterNum, elementRef) {
        if (map.has(gutterNum)) {
            map.get(gutterNum).push(elementRef);
        }
        else {
            map.set(gutterNum, [elementRef]);
        }
    }
    removedFromMap(map, gutterNum, elementRef) {
        const elements = map.get(gutterNum);
        elements.splice(elements.indexOf(elementRef), 1);
        if (elements.length === 0) {
            map.delete(gutterNum);
        }
    }
    static ngTemplateContextGuard(dir, ctx) {
        return true;
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: SplitGutterDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    /** @nocollapse */ static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.8", type: SplitGutterDirective, isStandalone: true, selector: "[asSplitGutter]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: SplitGutterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[asSplitGutter]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
