// angular-split/esm2022/lib/gutter/split-gutter-dynamic-injector.directive.mjs
import { Injector, Directive, Input, ViewContainerRef, TemplateRef } from '@angular/core';
import { GUTTER_NUM_TOKEN } from './gutter-num-token';
import * as i0 from "@angular/core";
/**
 * This directive allows creating a dynamic injector inside ngFor
 * with dynamic gutter num and expose the injector for ngTemplateOutlet usage
 */
export class SplitGutterDynamicInjectorDirective {
    set gutterNum(value) {
        this.vcr.clear();
        const injector = Injector.create({
            providers: [
                {
                    provide: GUTTER_NUM_TOKEN,
                    useValue: value,
                },
            ],
            parent: this.vcr.injector,
        });
        this.vcr.createEmbeddedView(this.templateRef, { $implicit: injector });
    }
    constructor(vcr, templateRef) {
        this.vcr = vcr;
        this.templateRef = templateRef;
    }
    static ngTemplateContextGuard(dir, ctx) {
        return true;
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: SplitGutterDynamicInjectorDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    /** @nocollapse */ static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.8", type: SplitGutterDynamicInjectorDirective, isStandalone: true, selector: "[asSplitGutterDynamicInjector]", inputs: { gutterNum: ["asSplitGutterDynamicInjector", "gutterNum"] }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: SplitGutterDynamicInjectorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[asSplitGutterDynamicInjector]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }], propDecorators: { gutterNum: [{
                type: Input,
                args: ['asSplitGutterDynamicInjector']
            }] } });
